File: linear/linear.cpp
#include <cassert>
#include <iostream>

template <typename T>
class Vector 
{
public:

    Vector()
    {
        arr = nullptr;
        size = 0;
        capacity = 0;
    }

    ~Vector() 
    {
        delete[] arr;
        size = 0;
        capacity = 0;
    }

    void add(T el)
    {
        if(size == capacity) {
            resize();
        }

        arr[size++] = el;
    }

    bool remove(int index)
    {
        if(index >= size || index < 0) {
            return false;
        } else {
            for (int i = index; i < size - 1; ++i) {
                arr[i] = arr[i + 1];
            }
            --size;
            return true;
        }
    }

    const T* get(int index) const
    {
        if(index >= size || index < 0) {
            return nullptr;
        } else {
            return &arr[index];
        }
    }

    int length() const
    {
        return size;
    }

    void debug()
    {
        for(int i = 0; i < size; ++i) {
            std::cout << arr[i] << " ";
        }
        std::cout << std::endl;
    }

private:

    void resize()
    {
        T* toDelete = arr;
        capacity = (capacity == 0 ? 1 : capacity) * 2;
        T* new_arr = new T[capacity];
        for(int i = 0; i < size; ++i) {
            new_arr[i] = arr[i];
        }
        arr = new_arr;
        delete[] toDelete;
    }

private:
    T* arr;
    int size;
    int capacity;
};

template <typename T>
class Stack: public Vector<T>
{

public:
    void push(T el)
    {
        Vector<T>::add(el);
    }

    void pop()
    {
        if(Vector<T>::length() != 0) {
            Vector<T>::remove(Vector<T>::length() - 1);
        }
    }

    const T* peek() const
    {
        if(Vector<T>::length() != 0) {
            return Vector<T>::get(Vector<T>::length() - 1);
        } else {
            return nullptr;
        }
    }
};

template <typename T>
class Queue: public Vector<T>
{
public:
    void enqueue(T el)
    {
        Vector<T>::add(el);
    }

    void dequeue()
    {
        if(Vector<T>::length() != 0) {
            Vector<T>::remove(0);
        }
    }

    const T* peek() const
    {
        if(Vector<T>::length() != 0) {
            return Vector<T>::get(0);
        } else {
            return nullptr;
        }
    }
    
};

template <typename T>
class CyclicQueue
{
    struct Node
    {
        T val;
        Node* next, *prev;
    };

public:

    CyclicQueue()
    {
        Node* initNode = new Node();
        initNode->next = initNode;
        initNode->prev = initNode;

        from = to = initNode;
        size = 0;
    }

    ~CyclicQueue()
    {
        del(from);
    }

    void enqueue(T el)
    {
        if (from == to->next) {
            resize();
        }
        
        to->val = el;
        to = to->next;
        ++size;
    }

    void dequeue()
    {
        if (from != nullptr && from != to) {
            from = from->next;
            --size;
        }

    }

    const T* peek() const
    {
        if (from != nullptr && from != to) {
            return &from->val;
        } else {
            return nullptr;
        }
    }

    int length() const
    {
        return size;
    }

private:

    // must only be called when the queue is full
    void resize()
    {
        int numToAdd = size == 0 ? 1 : size;
        // std::cout << "resizing " << size << std::endl;
        for(int i = 0; i < numToAdd; ++i) {
            addNode();
        }
        
    }

    void addNode()
    {
        Node* prevNode = to;
        Node* nextNode = to->next;

        Node* newNode = new Node();
        newNode->prev = prevNode;
        newNode->next = nextNode;
        nextNode->prev = newNode;
        prevNode->next = newNode;
    }

    void del(Node* toDelete)
    {
        del(toDelete->next);
        delete toDelete;
        toDelete = nullptr;
    }

private:
    Node* from, *to;
    int size;
};

// linked deque, priority queue
int main()
{
    // stack

    Stack<int>* myStack = new Stack<int>();
    assert(myStack->length() == 0);

    for(int i = 0; i < 10; ++i) {
        myStack->push(i);
        assert(myStack->length() == i + 1);
        assert(*(myStack->peek()) == i);
    }
    myStack->pop();
    assert(*(myStack->peek()) == 8);
    assert(myStack->length() == 9);
    
    // queue

    Queue<int>* myQueue = new Queue<int>();
    assert(myQueue->length() == 0);

    for(int i = 0; i < 10; ++i) {
        // myQueue->debug();
        myQueue->enqueue(i);
        assert(myQueue->length() == i + 1);
    }

    for(int i = 0; i < 10; ++i) {
        assert(*(myQueue->peek()) == i);
        myQueue->dequeue();
        assert(myQueue->length() == 10 - i - 1);
    }

    CyclicQueue<int>* myCyclicQueue = new CyclicQueue<int>();


    for(int i = 0; i < 10; ++i) {
        myCyclicQueue->enqueue(i);
        assert(myCyclicQueue->length() == i + 1);
    }

    for(int i = 0; i < 10; ++i) {
        assert(*(myCyclicQueue->peek()) == i);
        myCyclicQueue->dequeue();
        assert(myCyclicQueue->length() == 10 - i - 1);
    }

    return 0;
}

File: linear/linear.py
class Stack:

    def __init__(self):
        self.arr = []

    def push(self, el):
        self.arr.append(el)

    def pop(self):
        self.arr.pop(len(self.arr) - 1)

    def peek(self):
        return self.arr[len(self.arr) - 1]

    def length(self):
        return len(self.arr)
    
class Queue:

    def __init__(self):
        self.arr = []

    def enqueue(self, el):
        self.arr.append(el)

    def dequeue(self):
        self.arr.pop(0)

    def peek(self):
        return self.arr[0]

    def length(self):
        return len(self.arr)

class CyclicQueue:

    class Node:

        def __init__(self, value, prev_node, next_node):
            self.val = value
            self.prev = prev_node
            self.next = next_node

    def __init__(self):
        start = CyclicQueue.Node(None, None, None)
        start.prev = start.next = start
        self.frm = start
        self.to = start
        self.size = 0

    def enqueue(self, el):
        if self.to.next == self.frm:
            self.resize()

        self.to.val = el
        self.to = self.to.next
        self.size += 1

    def dequeue(self):
        if self.size != 0:
            self.frm = self.frm.next
            self.size -= 1

    def peek(self):
        if self.size != 0:
            return self.frm.val
        return None

    def length(self):
        return self.size

    def resize(self):
        prev_node = self.to
        next_node = self.to.next

        new_node = CyclicQueue.Node(None, prev_node, next_node)
        prev_node.next = new_node
        next_node.prev = new_node


my_stack = Stack()
assert(my_stack.length() == 0)

for i in range(10):
    my_stack.push(i);
    assert(my_stack.length() == i + 1)
    assert(my_stack.peek() == i)

my_stack.pop()
assert(my_stack.peek() == 8)
assert(my_stack.length() == 9)

# queue

my_queue = Queue();
assert(my_queue.length() == 0)

for i in range(10):
    my_queue.enqueue(i)
    assert(my_queue.length() == i + 1)


for i in range(10):
    assert(my_queue.peek() == i)
    my_queue.dequeue()
    assert(my_queue.length() == 10 - i - 1)


my_cyclic_queue = CyclicQueue()


for i in range(10):
    my_cyclic_queue.enqueue(i)
    assert(my_cyclic_queue.length() == i + 1)


for i in range(10):
    assert(my_cyclic_queue.peek() == i)
    my_cyclic_queue.dequeue()
    assert(my_cyclic_queue.length() == 10 - i - 1)



